### PR TITLE
feat: display extension icons

### PR DESCRIPTION
**Pull request by vzhukovs**
### What does this PR do?
Current changes proposal adds An ability to display extension icons.

Signed-off-by: Vladyslav Zhukovskyi <vzhukovs@redhat.com>

### Screenshot/screencast of this PR

Снимок экрана 2022-12-19 в 14 55 13

### What issues does this PR fix or reference?

#393 

### How to test this PR?

Try to install Docker Desktop Extension:

```
docker/disk-usage-extension:0.2.5
docker/logs-explorer-extension:0.2.1
```

### CODE REVIEW

**Comment by vzhukovs** [comment]
cc @slemeur

File: packages/main/src/plugin/contribution-manager.ts [hunk 1]
```diff
@@ -78,6 +81,26 @@ export class ContributionManager {
...
+  async loadBase64Icon(rootDirectory: string, metadata: any): Promise<string> {
+    if (!metadata.icon) {
+      //return an empty svg icon <svg xmlns="http://www.w3.org/2000/svg" width="1" height="1"/>
+      return 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxIiBoZWlnaHQ9IjEiLz4=';
+    }
+    const iconPath = path.join(rootDirectory, metadata.icon);
+    if (!fs.existsSync(iconPath)) {
+      throw new Error('Invalid icon path : ' + iconPath);
+    }
+    return new Promise((resolve, reject) => {
```
Comment (benoitf) [inline packages/main/src/plugin/contribution-manager.ts]: you can use fs.promises to have an async/await method to read file

**Comment by slemeur** [comment]
Thanks @vzhukovs !

It feels to me the that alignement of the extension name is inconsistent. 

unnamed

On the other side, it feels that the icon of the extensions all have a different size. Is there a way for us to improve this?

File: packages/main/src/plugin/contribution-manager.ts (hunk 1 above)
Comment (vzhukovs) [reply packages/main/src/plugin/contribution-manager.ts]: updated

**Comment by vzhukovs** [comment]
>On the other side, it feels that the icon of the extensions all have a different size. Is there a way for us to improve this?

Each icon is scaled to 24x24px size, and read directly from the each extension. Not sure if we can update for example positioning.

>It feels to me the that alignement of the extension name is inconsistent.

Will try to take a look, how we can align items in this list to one base line.

**Comment by cdrage** [comment]
off topic, but I think we should do the same for providers

**Comment by vzhukovs** [comment]
@benoitf @cdrage @slemeur  I've updated the PR, fixed issue with the icons aligning:
extension icons

File: packages/main/src/plugin/contribution-manager.ts [hunk 2]
```diff
@@ -78,6 +81,20 @@ export class ContributionManager {
     this.apiSender.send('contribution-register', this.contributions);
   }
 
+  async loadBase64Icon(rootDirectory: string, metadata: any): Promise<string> {
```
Comment (cdrage) [inline packages/main/src/plugin/contribution-manager.ts:89]: lint error, no `any` on metadata / something else on it?

**Review by cdrage — CHANGES_REQUESTED** [review]
1 small comment, otherwise LGTM. Integration works wonderfully, icons are loaded and changing the size of the window updates it / aligns it correctly.

File: packages/main/src/plugin/contribution-manager.ts (hunk 2 above)
Comment (vzhukovs) [reply packages/main/src/plugin/contribution-manager.ts:89]: `metadata` is a raw representation of metadata.json, I think, we can add
`// eslint-disable-next-line @typescript-eslint/no-explicit-any`
as it has been done for `loadMetadata` method, wdyt?

Comment (vzhukovs) [reply packages/main/src/plugin/contribution-manager.ts:89]: added a comment to the line

Comment (cdrage) [reply packages/main/src/plugin/contribution-manager.ts:89]: 👍

**Review by cdrage — APPROVED** [review]

**Comment by cdrage** [comment]
LGTM

File: packages/main/src/plugin/contribution-manager.ts [hunk 3]
```diff
@@ -78,6 +81,21 @@ export class ContributionManager {
     this.apiSender.send('contribution-register', this.contributions);
   }
 
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  async loadBase64Icon(rootDirectory: string, metadata: any): Promise<string> {
+    if (!metadata.icon) {
+      //return an empty svg icon <svg xmlns="http://www.w3.org/2000/svg" width="1" height="1"/>
+      return 'data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHdpZHRoPSIxIiBoZWlnaHQ9IjEiLz4=';
```
Comment (benoitf) [inline packages/main/src/plugin/contribution-manager.ts]: maybe a string constant like EMPTY_ICON

File: packages/renderer/src/AppNavigation.svelte [hunk 4]
```diff
@@ -49,7 +50,7 @@ export let meta;
 
 <svelte:window bind:innerWidth="{innerWidth}" />
 <nav
-  class="pf-c-nav z-0 group w-12 hover:w-[180px] md:w-[180px] md:min-w-[180px] shadow flex-col justify-between sm:flex transition-all duration-500 ease-in-out"
+  class="pf-c-nav z-0 group w-14 hover:w-[250px] md:w-[250px] hover:sm:w-[250px] md:min-w-[200px] shadow flex-col justify-between sm:flex transition-all duration-500 ease-in-out overflow-hidden hover:overflow-y-auto"
```
Comment (benoitf) [inline packages/renderer/src/AppNavigation.svelte:53]: It increases the width of the left navbar no ?

**Review by lstocchi — COMMENTED** [review]
Icons LGTM
On Windows the row doesn't take the full width. It would be better to have it
highlight_too_short

File: packages/main/src/plugin/contribution-manager.ts (hunk 3 above)
Comment (vzhukovs) [reply packages/main/src/plugin/contribution-manager.ts]: updated

Comment (vzhukovs) [reply packages/main/src/plugin/contribution-manager.ts]: Width of collapsed left bar was increased to have equal spacing on the icons:

<details>
  <summary>Before update:</summary>

Monosnap Podman Desktop 2023-02-09 12-42-10
</details>

<details>
  <summary>After setting width to 3.5rem:</summary>

Monosnap Podman Desktop 2023-02-09 12-37-22

</details>

Width of the left navigation bar was increased a bit to not to have issues with displaying extension list:

<details>
  <summary>Before update(default opening state):</summary>

Monosnap Podman Desktop 2023-02-09 12-48-03

</details>

<details>
  <summary>After setting new width values(default opening state):</summary>

Monosnap Podman Desktop 2023-02-09 12-39-55

</details>


<details>
  <summary>Before update(scaled window to 1300px):</summary>

Monosnap Podman Desktop 2023-02-09 12-46-40

</details>


<details>
  <summary>After update(scaled window to 1300px):</summary>

Monosnap Podman Desktop 2023-02-09 12-40-24

</details>

Visually, difference is little, but extension list looks prettier.

File: packages/renderer/src/AppNavigation.svelte (hunk 4 above)
Comment (vzhukovs) [reply packages/renderer/src/AppNavigation.svelte:53]: Width of collapsed left bar was increased to have equal spacing on the icons:

<details>
  <summary>Before update:</summary>

Monosnap Podman Desktop 2023-02-09 12-42-10
</details>

<details>
  <summary>After setting width to 3.5rem:</summary>

Monosnap Podman Desktop 2023-02-09 12-37-22

</details>

Width of the left navigation bar was increased a bit to not to have issues with displaying extension list:

<details>
  <summary>Before update(default opening state):</summary>

Monosnap Podman Desktop 2023-02-09 12-48-03

</details>

<details>
  <summary>After setting new width values(default opening state):</summary>

Monosnap Podman Desktop 2023-02-09 12-39-55

</details>


<details>
  <summary>Before update(scaled window to 1300px):</summary>

Monosnap Podman Desktop 2023-02-09 12-46-40

</details>


<details>
  <summary>After update(scaled window to 1300px):</summary>

Monosnap Podman Desktop 2023-02-09 12-40-24

</details>

Visually, difference is little, but extension list looks prettier.

**Comment by vzhukovs** [comment]
@lstocchi I've fixed the problem with displaying items on windows, now it looks better:

Podman+Desktop+2023-02-09+13 32 21

**Review by lstocchi — APPROVED** [review]
It looks fine. Thanks!! 🚀 

full_width_icon

**Comment by slemeur** [comment]
Good work @vzhukovs ! LGTM

**Pull request merged**